alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs

```diff
@@ -22,56 +35,82 @@ import Cardano.Ledger.Alonzo.TxBody
 import qualified Plutus.V1.Ledger.Crypto as P (PubKeyHash (..))
 import qualified Plutus.V1.Ledger.Interval as P
   ( Extended (..),
     Interval (..),
     LowerBound (..),
     UpperBound (..),
   )
-import qualified Plutus.V1.Ledger.Scripts as P (DatumHash (..), ValidatorHash (..))
+import qualified Plutus.V1.Ledger.Scripts as P (Datum (..), DatumHash (..), ValidatorHash (..))
 import qualified Plutus.V1.Ledger.Slot as P (SlotRange)
-import qualified Plutus.V1.Ledger.Tx as P (TxOut (..), TxOutRef (..), TxOutType (..))
+import qualified Plutus.V1.Ledger.Tx as P (TxOutRef (..))
 import qualified Plutus.V1.Ledger.TxId as P (TxId (..))
 import qualified Plutus.V1.Ledger.Value as P (CurrencySymbol (..), TokenName (..), Value (..), singleton, unionWith)
-import Shelley.Spec.Ledger.Address (Addr (..), BootstrapAddress (..))
+import Shelley.Spec.Ledger.Address (Addr (..))
 import Shelley.Spec.Ledger.BaseTypes (StrictMaybe (..))
-import Shelley.Spec.Ledger.Credential (Credential (KeyHashObj, ScriptHashObj))
-import Shelley.Spec.Ledger.Keys (KeyHash (..))
+import Shelley.Spec.Ledger.Credential (Credential (KeyHashObj, ScriptHashObj), Ptr (..), StakeReference (..))
+import Shelley.Spec.Ledger.Keys (KeyHash (..), KeyRole (..), hashKey)
 import Shelley.Spec.Ledger.Scripts (ScriptHash (..))
-import Shelley.Spec.Ledger.TxBody (TxId (..), TxIn (..))
+import Shelley.Spec.Ledger.TxBody (TxId (..), TxIn (..), WitVKey (..))
 import Shelley.Spec.Ledger.UTxO (UTxO (..))
 
 -- =========================================================
 
+-- | The _ptrs field of the DState (State of staking pool delegations and rewards)
+type PtrMap crypto = Map.Map Ptr (Credential 'Staking crypto)
+
+transDataHash :: StrictMaybe (DataHash c) -> Maybe P.DatumHash
+transDataHash (SJust safe) = Just (transDataHash' safe)
+transDataHash SNothing = Nothing
+
+transDataHash' :: (DataHash c) -> P.DatumHash
+transDataHash' safe = (P.DatumHash (transSafeHash safe))
+
+transKeyHash :: KeyHash d c -> P.PubKeyHash
+transKeyHash (KeyHash (UnsafeHash h)) = P.PubKeyHash (fromShort h)
+
+transScriptHash :: ScriptHash i -> P.ValidatorHash
+transScriptHash (ScriptHash (UnsafeHash h)) = P.ValidatorHash (fromShort h)
+
 transSafeHash :: SafeHash c i -> BS.ByteString
 transSafeHash safe = case extractHash safe of UnsafeHash b -> fromShort b
 
 transTxId :: TxId era -> P.TxId
 transTxId (TxId safe) = P.TxId (transSafeHash safe)
 
+-- =====================
+
+transStakeCred :: Credential keyrole crypto -> BS.ByteString
+transStakeCred (ScriptHashObj (ScriptHash (UnsafeHash kh))) = (fromShort kh)
+transStakeCred (KeyHashObj (KeyHash (UnsafeHash kh))) = (fromShort kh)
+
+transStakeReference :: (PtrMap crypto) -> StakeReference crypto -> Maybe P.StakingHash
+transStakeReference _mp (StakeRefBase cred) = Just (P.StakingHash (transStakeCred cred))
+transStakeReference mp (StakeRefPtr ptr) =
+  case Map.lookup ptr mp of
+    Nothing -> Nothing
```

Comment:
Is this impossible if we've got to this stage of validation? Does the ledger allow addresses with dangling staking references? If it's impossible, could we make this fail validation instead in a more aggressive way? I'm not sure it will matter in practice, it just feels itchy to not be able to see the difference between this invariant failing and plain old `StakeRefNull`.

Reply:
>  Does the ledger allow addresses with dangling staking references?

yes, it does. someone can place lovelace in a pointer address and then de-register their stake credential, leaving a dangling staking references. so we don't even bother to enforce valid pointers on creation of an output. Addresses with dangling pointers behave identically like `StakeRefNull`. But if plutus would like to make such a distinction, we could pass some `Dangling` type to the interpreter.

Reply:
Hmmmmmm. I don't *think* so. I'm trying to think of edge cases, but I think it's fine. Worth a comment, perhaps.

Reply:
sure, comments are always great. but I'm honestly happy either way. we keep it as is unless you think of something.

Reply:
we discussed this in person, but I'll make a note here. we are _not_ going to resolve pointer addresses. that would break determinism.

Reply:
We now pas in pointers, but we do not resolve them, so you can check if two pointers are the same, but you can't tell what they point to.